doc/source/whatsnew/v1.5.0.rst

```diff
@@ -820,6 +820,7 @@ I/O
 - Bug in :func:`read_parquet` when ``engine="fastparquet"`` where the file was not closed on error (:issue:`46555`)
 - :meth:`to_html` now excludes the ``border`` attribute from ``<table>`` elements when ``border`` keyword is set to ``False``.
 - Bug in :func:`read_sas` returned ``None`` rather than an empty DataFrame for SAS7BDAT files with zero rows (:issue:`18198`)
+- Bug in :meth:`TextReader.read` with specified (non-extension) integer ``dtype`` can cause silent overflow or unexpected return dtype (:issue:`47167`)
```

Comment:
`TextReader` is not in the api docs, so I suspect the link will not render. can you describe the issue in terms of the public api.

Reply:
Most prominently this affects :func:`read_csv` when `engine="c"`. Honestly, I forgot to try this for `engine="python"` and I now recognize that the same thing happens there: <https://github.com/pandas-dev/pandas/blob/aa305f3608938b5bfd9d08c75ee8b6d883f7098b/pandas/io/parsers/base_parser.py#L786-L791> We could introduce the same check like for the c-parser (while excluding the cases where `(a == b).all()` is not necessary to check):
```python
try:
    casted = astype_nansafe(values, cast_type, copy=True, skipna=True)
except ValueError as err:
    # unchanged ...
if (
    is_integer_dtype(cast_type)
    and (pandas_dtype(cast_type) != np.uint64)
    and (not (casted == values).all())
):
    raise TypeError(
        f"cannot safely cast non-equivalent {values.dtype} to {cast_type}"
    )
values = casted
```
I am usure if it might be even better to check for int-overflows directly in `astype_nansafe`, where many dtype-specific cases already exist. I am afraid to touch this method, since it is heavily used. This would also extend the scope of int-overflow checks beyond parsing to casting in general (maybe with an optional parameter like `raise_on_overflow` that is False by default). If you don't encourage me to go this way, I'd rather include the check in base_parser.py as described above. What's your opinion?